File: src/ruby/nodes/BinaryExpr.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'treetop'
require_relative 'BooleanExpr'

# Variable interpolation in a template
class BinaryExpr < BooleanExpr
  def truthy?(context)
    case op.text_value
    when 'and'
      lhs.truthy?(context) && rhs.truthy?(context)
    when 'or'
      lhs.truthy?(context) || rhs.truthy?(context)
    end
  end
end

File: cpp/src/nodes/binary_expr.cpp
#include "binary_expr.hpp"

const std::string BinaryExpr::AND_STR = "and";
const std::string BinaryExpr::OR_STR = "or";

BinaryExpr::BinaryExpr(BooleanPtr lhs, Operator op, BooleanPtr rhs):
  lhs(lhs),
  op(op),
  rhs(rhs)
{}

bool BinaryExpr::truthy(Json &context) const {
  switch (op) {
    case AND:
      return lhs->truthy(context) && rhs->truthy(context);
    case OR:
      return lhs->truthy(context) || rhs->truthy(context);
  }
}
